File: week01/한정민/PGS_스티커 모으기(2).py
def solution(sticker):
    if len(sticker) == 1:
        return sticker[0]
        
    dp1 = [0] * len(sticker)
    dp1[0] = sticker[0]
    dp1[1] = sticker[0]
    for i in range(2, len(sticker) - 1):
        dp1[i] = max(dp1[i-1], dp1[i-2] + sticker[i])
    
    dp2 = [0] * len(sticker)
    dp2[1] = sticker[1]
    for i in range(2, len(sticker)):
        dp2[i] = max(dp2[i-1], dp2[i-2] + sticker[i])
            
    return max(dp1[-2], dp2[-1])

File: week01/이채은/PGS_스티커 모으기(2).java
import java.util.*;

class Solution {
    public int solution(int sticker[]) {
        int N = sticker.length;

        if(N == 1) return sticker[0];
        else if(N == 2) return Math.max(sticker[0], sticker[1]);

        //0 ~ N-1
        int[] dp = new int[N-1];
        dp[0] = sticker[0];
        dp[1] = Math.max(sticker[0],sticker[1]);
        for(int i = 2; i < N-1; i++){
            dp[i] = Math.max(dp[i-1], sticker[i] + dp[i-2]);
        }

        //1 ~ N
        int[] dp2 = new int[N-1];
        dp2[0] = sticker[1];
        dp2[1] = Math.max(sticker[1],sticker[2]);
        for(int i = 2; i < N-1; i++){
            dp2[i] = Math.max(dp2[i-1], sticker[i+1] + dp2[i-2]);
        }

        return Math.max(dp[N-2], dp2[N-2]);
    }
}